### ISSUE TITLE
Extending Copilot with the CDK

### ISSUE THREAD

**Issue by efekarakus** [aws/copilot-cli#4208]
Hi folks! This issue is a design proposal for overriding Copilot abstracted resources using the [AWS Cloud Development Kit (CDK)](https://docs.aws.amazon.com/cdk/v2/guide/home.html). The goal is to provide a "break the glass" mechanism to access and configure functionality that is not surfaced by [Copilot manifests](https://aws.github.io/copilot-cli/docs/manifest/overview/) by leveraging the expressive power of a programming language.

### Problem statement
Clients can describe their application configuration with declarative [`manifest.yml` files](https://aws.github.io/copilot-cli/docs/manifest/lb-web-service/). Manifests enable users to define their application in terms of [“architecture as code”](https://aws.amazon.com/blogs/containers/containers-and-infrastructure-as-code-like-peanut-butter-and-jelly/) that gets transformed into AWS CloudFormation (CFN) templates by Copilot. Today, customers have the ability to define additional AWS resources with [“addons”](https://aws.github.io/copilot-cli/docs/developing/addons/modeling/) CFN templates. However, the resources encapsulated by manifests remain unmodifiable by users. We're looking into providing extension solutions for customers to self-unblock themselves and access the internal layers of Copilot.

This document proposes a user experience for extending manifests with the CDK.
### Proposal
1. **Copilot will introduce a new command `copilot [noun] override`** to create a CDK application that overrides Copilot resources, for example:
   ```console
   $ copilot svc override
   > Which service's resources would you like to override?
   * frontend
   * api

   > Which infrastructure as code tool would you like to use to override "frontend"?
   * AWS Cloud Development Kit (CDK)
   * CloudFormation YAML patches

   > Which programming language would you like to use with the CDK?
   * Typescript(ts)

   > Which resources in "frontend" would you like to to override?
   [ ] DiscoveryService          (AWS::ServiceDiscovery::Service)
   [x] Service                   (AWS::ECS::Service)
   [x] PublicNetworkLoadBalancer  (AWS::ElasticLoadBalancingV2::LoadBalancer)
   [ ] TaskDefinition            (AWS::ECS::TaskDefinition)
   ✅ Created a new CDK app under `copilot/frontend/overrides/` to override the resources:
   .
   ├── .gitignore
   ├── .build/
   ├── bin/
   │   └── override.ts
   ├── cdk.json
   ├── package.json
   ├── stack.ts
   ├── tsconfig.json
   └── README.md

   Copilot did not detect npm installed.
   > Would you like Copilot to install npm? [Y/n] y
   ✅ Installed npm v9.1.1 and NodeJS v1.18.2.
   ✅ Ran `npm install` to install dependencies.

   > Recommended Actions:
   1. Please follow the guide under README.md
   ```
   The file structure generated by the `override` command is  equivalent to `cdk init` output. The only additional file created by Copilot is the hidden `.build/` directory. 

   In order to simplify onboarding with the CDK, Copilot will assist clients with installing their dependencies. On macOS and Linux, Copilot can install `npm` for customers on their local machine. On Windows, we won't be able to help customers with installing `npm` and will skip to asking them to follow the README instead. Once `npm` is present, Copilot will run `npm install` to install the dependencies on behalf of the client.

2. **Clients have to edit only `stack.ts` file** in order to implement the `transform*` methods for the selected resources. The following content will be present in the file by default:
   ```ts
   import * as cdk from 'aws-cdk-lib';
   import { aws_elasticloadbalancingv2 as elbv2 } from 'aws-cdk-lib';
   import { aws_ecs as ecs } from 'aws-cdk-lib';

   export class TransformedStack extends cdk.Stack {
       constructor (scope: cdk.App, id: string, props?: cdk.StackProps) {
           super(scope, id, props);
            this.template = new cdk.cloudformation_include.CfnInclude(this, 'Template', {
               templateFile: path.join('.build', 'in.yaml'),
           });
           this.appName = template.getParameter('AppName').valueAsString;
           this.envName = template.getParameter('EnvName').valueAsString;
   
           this.transformService();
           this.transformPublicNetworkLoadBalancer();
       }
    
       // TODO: implement me.
       transformService() {
         const service = this.template.getResource("Service") as ecs.CfnService;
         throw new error("not implemented");
       }

       // TODO: implement me.
       transformPublicNetworkLoadBalancer() {
         const publicNetworkLoadBalancer = this.template.getResource("PublicNetworkLoadBalancer") as elbv2.CfnLoadBalancer;
         throw new error("not implemented");
       }
   }
   ```
   As can be seen above, Copilot will use the [clouformation_include module](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.cloudformation_include-readme.html)  provided by the CDK to help author transformations. This library is the CDK’s [recommendation](https://docs.aws.amazon.com/cdk/v2/guide/use_cfn_template.html)  from their “Import or migrate an existing AWS CloudFormation template” guide. It enables accessing the resources hidden by the manifest as L1 constructs and ensures the CDK retains the resources on synthesis. The `CfnInclude` object is initialized from the `.build/in.yaml` CFN template. This is how Copilot and the CDK communicates. Copilot writes the manifest generated CFN template under the `.build/` directory, which then gets parsed by the `cloudformation_include` library into a CDK construct.

That's it! From this point forward, `copilot svc package` or `copilot svc deploy` will apply the overrides written in `stack.ts`. Under the hood, Copilot will invoke `cdk synth` to synthesize the transformed template and use that to deploy to the existing CFN stack.
### Feedback
We'd love to hear your feedback on the user experience above:
1. Do you see yourself using this feature? (overriding resources with the CDK)
2. Does the feature meet your needs? 
3. If not, can you elaborate on what you'd like to do that can't be done with the above proposal?

### Appendix: Sample issues
Every new “capability” feature request can be boiled down to having an override functionality. I’ve captured only a sample of these to test the proposal against.

|Link	|Description	|
|---	|---	|
|[#4063](https://github.com/aws/copilot-cli/issues/4063)	|Override `nlb` configuration to assign elastic IPs to the network load balancer	|
|[#4010](https://github.com/aws/copilot-cli/issues/4010)	|Assign `ReadOnlyRootFS: true` to all containers in the task definition by default	|
|[#4005](https://github.com/aws/copilot-cli/issues/4005), [#3840](https://github.com/aws/copilot-cli/issues/3840)	|Support for internal NLB for Backend Services. Customers can create the nlb with addons/ but need to wire it by overriding the `ECS::Service` definition	|
|[#3733](https://github.com/aws/copilot-cli/issues/3733)	|Allow setting external launch type for ECS tasks	|
|[#3721](https://github.com/aws/copilot-cli/issues/3721)	|Allow setting VPC flow logs for environments	|
|[#3720](https://github.com/aws/copilot-cli/issues/3720)	|Turn off assigning public IP addresses for tasks launched in public subnets	|
|[#3594](https://github.com/aws/copilot-cli/issues/3594), [#1783](https://github.com/aws/copilot-cli/issues/1783)	|Multiple port support for services behind NLB and ALB	|
|[#3506](https://github.com/aws/copilot-cli/issues/3506)	|Add additional permissions to the generated environment manager role	|
|[#3504](https://github.com/aws/copilot-cli/issues/3504)	|Specify additional security groups for the app runner vpc connector	|

**Comment by efekarakus** [comment]
For an alternative proposal that uses simple CloudFormation YAML patches, see https://github.com/aws/copilot-cli/issues/4209!

**Comment by RichiCoder1** [comment]
I really love this! The only addition I'd have is to take it even further. A lot of the power of the CDK is high level abstractions, so it'd be nice to have a Copilot abstraction that takes care of details like knowing where/how to template include and possibly having opinionated/type-safe ways of getting templated resources.

**Comment by efekarakus** [comment]
Hi @RichiCoder1 👋 

Thanks for the feedback! Are you thinking of something along these lines:
```ts
import * as copilot from 'aws-copilot-lib';

const props = copilot.buildPropsFromManifest('../manifest.yml');
const lbws = new copilot.LoadBalancedWebService(this, 'Web', props);
const service = lbws.getService(); // returns L2 ECS service construct.                  
```

One benefit that I can think of with this route is that it would allow you to compose CDK apps made out of Copilot abstractions 👍. 
Or were you thinking of a different option?

-----

>  that takes care of details like knowing where/how to template include

For clarification, when running `copilot svc/env override`, Copilot will scaffold the following piece of code:
```ts
constructor (scope: cdk.App, id: string, props?: cdk.StackProps) {
      super(scope, id, props);
       this.template = new cdk.cloudformation_include.CfnInclude(this, 'Template', {
          templateFile: path.join('.build', 'in.yaml'), 
      });
```
And everytime `copilot svc package` or `copilot deploy` is run Copilot will write the CloudFormation template automatically under `.build/in.yaml` so _no user action is needed_.

> opinionated/type-safe ways of getting templated resources

Similarly, Copilot will scaffold the following line:
```ts
const service = this.template.getResource("Service") as ecs.CfnService;
```
The service variable is now a type-safe **L1** CDK construct: https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_ecs.CfnService.html

**Do these responses mitigate your concerns?**

**Comment by benjaminpottier** [comment]
I think this is great! I have a few questions/comments:

- Will there be a beta we can use before the final release of this feature (if it is released)?
- Currently, any generated cloudformation from copilot isn't visible until its actually deployed to AWS. Would there be a synth-like command that could be run from copilot to show us everything that will be deployed (not just the CDK we've written to override/expand services/jobs etc.)
- I think having constructs that represents the abstracted "service" or "job" etc would be very helpful. That way we can work with a single construct as opposed to working individually with each resource that composes the abstraction/pattern. See: https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_ecs_patterns-readme.html ... I'm not sure of the patterns here track with what copilot releases.

**Comment by Lou1415926** [comment]
Hello @benjaminpottier! 

I can probably answer your second question - inviting @efekarakus for the other two!

> Currently, any generated cloudformation from copilot isn't visible until its actually deployed to AWS. Would there be a synth-like command that could be run from copilot to show us everything that will be deployed (not just the CDK we've written to override/expand services/jobs etc.)

You should be able to run `copilot svc package` to preview the CFN template that would be generated after applying the CDK code changes.  In addition, we are thinking about adding  `--diff` flag to `copilot svc package` for folks to view the difference between the deployed template and the local template. Would these be helpful for you?

**Comment by bpottier** [comment]
> Hello @benjaminpottier!
> 
> I can probably answer your second question - inviting @efekarakus for the other two!
> 
> > Currently, any generated cloudformation from copilot isn't visible until its actually deployed to AWS. Would there be a synth-like command that could be run from copilot to show us everything that will be deployed (not just the CDK we've written to override/expand services/jobs etc.)
> 
> You should be able to run `copilot svc package` to preview the CFN template that would be generated after applying the CDK code changes. In addition, we are thinking about adding `--diff` flag to `copilot svc package` for folks to view the difference between the deployed template and the local template. Would these be helpful for you?

That answers my question. Thank you! And yes, diff would be helpful.

**Comment by craigjbass** [comment]
We are currently planning to use copilot in "compliance environments" (UK Public Sector), where individual development teams have limited access to AWS APIs (i.e. no ability to apply CF from local terminal).

What this looks like is using Copilot strategically as a tool as part of a larger "platform offering" from a centralised platform team. I do wonder how these advanced override features will evolve over time and be suitable in those scenarios. 

My current view is that there would either need to be an abstraction around copilot, or some sort of copilot-managed concept that groups "Applications" into a wider platform, and allows cross-cutting changes to be applied to many teams copilot managed services.

**Comment by efekarakus** [comment]
Hi @bpottier, @craigjbass ! 
Apologies on the late reply, I was at re:invent 😎 

> Will there be a beta we can use before the final release of this feature (if it is released)?

We can happily share a pre-release binary here before submission to get early feedback on the feature, that's a great idea 👍 

----

> I think having constructs that represents the abstracted "service" or "job" etc would be very helpful. 

That's really interesting, we can definitely create a L3 construct that takes the same manifest file as an object as input. Similar to in [this comment](https://github.com/aws/copilot-cli/issues/4208#issuecomment-1325420230):
```ts
const lbws = new copilot.LoadBalancedWebService(this, 'Web', {
  http: {
     path: '/',
     targetPort: 80,
  },
  image: {
     build: './Dockerfile',
  },
});
```


**Are you inclined for this abstraction so that you can have all your infrastructure defined in the CDK?**

<details>
  <summary> Perhaps that would be a better fit for a <code>copilot eject</code> command </summary>

I'm guessing that the Copilot CDK construct would still be interoperable with Copilot commands like `copilot svc logs`.
*Replacing* the infrastructure definition from Copilot manifests to the CDK seems to be more around "ejecting" rather than extending Copilot. Which I think is understandable, the `aws-copilot-lib` library would be used to replace manifests and interoperate with the CDK, while Copilot the CLI can be used for operational commands.
</details>

-----

> My current view is that there would either need to be an abstraction around copilot, or some sort of copilot-managed concept that groups "Applications" into a wider platform, **and allows cross-cutting changes** to be applied to many teams copilot managed services.

Interesting! Can I ask some clarification questions @craigjbass 🥺:
1. Do you deem it's okay if this feature is reserved to the _platform_ team rather than the _dev_ teams?
2. Out of curiosity: How will _dev_ teams make changes to their services? Do they have access to the `manifest.yml` files and can do a git push to trigger the `copilot pipeline` to apply their changes?
3.Is the concern applying the same "extension" across multiple manifest files? If so I think a pattern like this can be possible:
   ```ts
   import {customTransform} from '../../common/transform-copilot.ts'; // Import from a common path or a published node module.
   
   constructor (scope: cdk.App, id: string, props?: cdk.StackProps) {
      super(scope, id, props);
       this.template = new cdk.cloudformation_include.CfnInclude(this, 'Template', {
          templateFile: path.join('.build', 'in.yaml'), 
      });
      customTransform(this.template.getResource("Service");  // Apply the transformation here.
   }
   ```

**Comment by craigjbass** [comment]
> 1. Do you deem it's okay if this feature is reserved to the platform team rather than the dev teams?

Yes, although I wonder how we'd enforce that without simply not providing the Copilot CLI to teams. I can see a benefit in giving some but not all of the functionality of the Copilot CLI directly to dev teams. (Mostly around operability)

> 2. Out of curiosity: How will dev teams make changes to their services? Do they have access to the manifest.yml files and can do a git push to trigger the copilot pipeline to apply their changes?

Unclear yet. Ideally, they would just use the copilot manifest, but if we can't put in the relevant controls - they will probably commit a manifest that a custom tool will read to produce the copilot manifest. Examples of areas where this might be beneficial are for applying security groups without referencing a security group by ID. 

> 3.Is the concern applying the same "extension" across multiple manifest files? If so I think a pattern like this can be possible:

Yes but also in restricting the scope of the contract between platform and dev teams.

**Comment by efekarakus** [comment]
> I can see a benefit in giving some but not all of the functionality of the Copilot CLI directly to dev teams. (Mostly around operability)

That totally makes sense.
For operability, if we are talking about commands such as `copilot svc show`, `copilot svc status`, `copilot svc logs` then I think we can get away by assigning dev teams IAM roles that have only read actions allowed and these commands should work.

If we want dev teams to be able to modify infrastructure, but only limited to particular actions and resources, I think we need to put some more thought into it. 💭 
Today, we allow [IAM role boundaries](https://aws.github.io/copilot-cli/blogs/release-v122/#iam-role-permissions-boundary) to ensure certain actions are denied but to me it sounds like instead we need ways of customizing the [CloudFormationExecutionRole](https://github.com/aws/copilot-cli/blob/mainline/internal/pkg/template/templates/environment/partials/cfn-execution-role.yml) to limit it only for devs. This way if the CFNExecRole is limited then devs having access to manifests and extensions won't matter

**Comment by craigjbass** [comment]
IAM Role boundaries look good - although it doesn't seem possible to apply that to an existing app?

**Comment by craigjbass** [comment]
In terms of bringing this back on topic to CDK extensions.

> If we want dev teams to be able to modify infrastructure, but only limited to particular actions and resources, I think we need to put some more thought into it. thought_balloon

Looking at this as a "Platform in a Box" capability of Copilot, what would be ideal is if Copilot allowed a central platform team to define templated "bits of infrastructure" which can be turned on and off inside the manifest yaml with something as simple as a boolean flag. Cross-cutting definitions of infrastructure, rather than giving dev teams the keys to AWS itself.

**Comment by efekarakus** [comment]
> IAM Role boundaries look good - although it doesn't seem possible to apply that to an existing app?

That is the case indeed today.

> Looking at this as a "Platform in a Box" capability of Copilot, what would be ideal is if Copilot allowed a central platform team to define templated "bits of infrastructure" which can be turned on and off inside the manifest yaml 

👍 We have plans in the future, using CDK extensions, for platform teams to define their own "Custom" manifests. Something like:
```
$ copilot svc init
Which service type best represents your service's architecture?
  > Load Balanced Web Service   (Internet to ECS on Fargate)
  > Backend Service             (ECS on Fargate)
  > Worker Service              (Events to SQS to ECS on Fargate)
  > Custom                      (Define your own using `copilot svc override`)
```
If I'm understanding correctly that seems to fit into the use case you're describing. Our plan is to  allow platform engineers to register a custom `type` with Copilot which can then be used by dev teams. The devs would interact with the manifest file, and platform teams using the CDK would define the transformations from manifest to infrastructure.

_Side note: have you explored [AWS Proton](https://docs.aws.amazon.com/proton/latest/userguide/Welcome.html)? They're specializing specifically for this usecase. They provide some advanced functionality too for platform teams such as monitoring the [version of the services](https://docs.aws.amazon.com/proton/latest/userguide/ag-template-versions.html) that are deployed across environments. If you have any questions for them, you can post [here](https://github.com/aws/aws-proton-public-roadmap/discussions/categories/q-a)._

**Comment by dannyrandall** [comment]
Hey all! I just added a design proposal to add a `--diff` flag to `deploy` and `package` that should be helpful when working with extensions. Feel free to take a look and leave any feedback over there! #4259

**Comment by RichiCoder1** [comment]
Apologies for the very late reply, wanted to follow up!

> One benefit that I can think of with this route is that it would allow you to compose CDK apps made out of Copilot abstractions 👍.

That would be interesting, but would also likely conflict with the pre-existing [ECS Patterns](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_ecs_patterns-readme.html) and [ECS Service Extensions](https://github.com/cdklabs/cdk-ecs-service-extensions#readme). I could see still creating them as an eject option or slightly better maintained alternative however.

> Similarly, Copilot will scaffold the following line:
> `const service = this.template.getResource("Service") as ecs.CfnService;`
> The service variable is now a type-safe L1 CDK construct: [docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_ecs.CfnService.html](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_ecs.CfnService.html)

My thought process was that since Copilot (roughly) knows the shape of the template, instead of using raw keys and casting to the correct L1, copilot could expose an abstraction that surfaces these L1's as properties.

Something like:

```typescript
constructor (scope: cdk.App, id: string, props?: cdk.StackProps) {
   super(scope, id, props);
   // This construct already knows where the template lives, but you can optionally override it.
   this.generated = new copilot.GeneratedTemplate(this, "Template" /*, { ... } */);
   // template.service is an L1 Service
   customTransform(this.genereted.service);
   // or to work with the raw bits
   this.generated.template
}
```

---

With that above stated I like the above ideas where while CDK handles the meat of the synth and deployment, Copilot still provides value via developer niceties like log tailing, debug assistance, etc...

**Comment by efekarakus** [comment]
> Instead of using raw keys and casting to the correct L1, copilot could expose an abstraction that surfaces these L1's as properties.

That makes a lot of sense! It reminds me of how Amplify overrides work ([example](https://docs.amplify.aws/cli/auth/override/))

**Comment by PedroAlvarado** [comment]
Here is what I'd love to see:

<img width="949" alt="image" src="https://user-images.githubusercontent.com/1258134/207206845-1ef183ee-4122-4d81-b748-9e6b6f9a2d15.png">

**Comment by efekarakus** [comment]
Thanks @PedroAlvarado for the awesome diagram 😍 ! I have a few questions if you don't mind 🙇 

1. Can you expand a little bit on the L3 builder interface? Do you have a sample pseudo-code to illustrate the interface in mind?
2. Do you not feel like the proposed solution above help you achieve (1) enforcing requirements and (2) enabling customization and new resources?

> Overrides for, say, LocalStack

This is really interesting, because it makes it sound like you would like to _customize the CLI itself_, say by changing the UI or commands, (like [here](https://github.com/aws/copilot-cli/commit/d7252b13bba933c223180ea201fc3d5006839b74)) and **not just the infrastructure**. 
3. Am I understanding correctly?

**Comment by efekarakus** [comment]
Hi folks!!

I'm really excited to say that AWS Copilot v1.27 is out and CDK extensions is now supported! 🚀 

* Guide: https://aws.github.io/copilot-cli/docs/developing/overrides/cdk/
* Blog post: https://aws.github.io/copilot-cli/blogs/release-v127/
* Release notes: https://github.com/aws/copilot-cli/releases/tag/v1.27.0

I'll close the issue now! If you have any questions, feedback, or suggestions please feel free to open new GitHub issues 🥳
